compiler/rustc_lint/src/types.rs

```diff
@@ -114,35 +115,14 @@ declare_lint! {
     "detects enums with widely varying variant sizes"
 }
 
-declare_lint! {
-    /// The `invalid_nan_comparisons` lint checks comparison with `f32::NAN` or `f64::NAN`
-    /// as one of the operand.
-    ///
-    /// ### Example
-    ///
-    /// ```rust
-    /// let a = 2.3f32;
-    /// if a == f32::NAN {}
-    /// ```
-    ///
-    /// {{produces}}
-    ///
-    /// ### Explanation
-    ///
-    /// NaN does not compare meaningfully to anything – not
-    /// even itself – so those comparisons are always false.
-    INVALID_NAN_COMPARISONS,
-    Warn,
-    "detects invalid floating point NaN comparisons"
-}
-
 #[derive(Copy, Clone)]
 pub struct TypeLimits {
     /// Id of the last visited negated expression
     negated_expr_id: Option<hir::HirId>,
 }
 
-impl_lint_pass!(TypeLimits => [UNUSED_COMPARISONS, OVERFLOWING_LITERALS, INVALID_NAN_COMPARISONS]);
+// This pass also emits `INVALID_NAN_COMPARISONS` but adding that to the list here leads to ICEs.
+impl_lint_pass!(TypeLimits => [UNUSED_COMPARISONS, OVERFLOWING_LITERALS]);
```

Comment:
Sadly this macro does not document what that list of lints actually *means*. So maybe it is a problem that the pass emits other lints as well, or maybe that's totally fine? Is it used for optimizations, to skip the pass if all lints are allowed? That would be a problem. I have no way to know.

Reply:
> Is it used for optimizations, to skip the pass if all lints are allowed?

This currently isn't done but will be in the future. #106983

This macro only generates a method named `get_lints`, afaik.

Reply:
> This macro only generates a method named get_lints, afaik.

And what does that method do?^^

> This currently isn't done but will be in the future. https://github.com/rust-lang/rust/issues/106983 (Also https://github.com/rust-lang/rust/pull/114026)

That is quite concerning.

>  What ICE is it?

It complains that the same lint got registered twice, or something like that.

Reply:
> And what does that method do?^^

Well, returns the list the macro takes - for what it's used for: https://github.com/rust-lang/rust/blob/13e6f24b9adda67852fb86538541adaa68aff6e8/compiler/rustc_lint/src/lib.rs#L266-L269

The macro that defines this combined lint pass is here:

https://github.com/rust-lang/rust/blob/13e6f24b9adda67852fb86538541adaa68aff6e8/compiler/rustc_lint/src/lib.rs#L187C8-L187C8

So the issue is that it was already in the `HardwiredLints` pass but you added it to `TypeLimits`. It's then added twice in `register_internals`. The logic within `TypeLimits` for emitting this lint should be moved to the `HardwiredLints` pass, or the other way around.

Reply:
Seems like currently it is just used as a convenience to have the lints registered, so it's not a big deal that some lint is missing from the list.

It makes no sense to move the logic to another pass IMO. I just moved the lint declaration because the lint is now also getting emitted from rustc_mir_build which does not depend on rustc_lints.